decompiler/pipeline/dataflowanalysis/common_subexpression_elimination.py

```diff
@@ -265,18 +261,31 @@ def _is_invalid_dominator(self, basic_block: BasicBlock, expression: Expression)
         usages_in_the_same_block = [usage for usage in self._usages[expression] if usage.block == basic_block]
         return any([isinstance(usage.instruction, Phi) for usage in usages_in_the_same_block])
 
-    def _insert_definition(self, definition: CfgInstruction):
+    def _insert_definition(self, instruction: Instruction, block: BasicBlock, index: int):
         """Insert a new intermediate definition for the given expression at the given location."""
-        definition.block.instructions.insert(definition.index, definition.instruction)
-        for subexpression in _subexpression_dfs(definition.instruction):
-            self._usages[subexpression][definition] += 1
+        block.instructions.insert(index, instruction)
+
+        def update_location(location: CfgInstructionLocation) -> CfgInstructionLocation:
+            if location.block == block and location.index < index:
+                return dataclasses.replace(location, index=location.index + 1)
+            else:
+                return location
+
+        # update positions of expression usages
+        for occurrences in self._usages.values():
+            for location in list(occurrences):
+                if location.block == block and location.index >= index:
+                    occurrences[dataclasses.replace(location, index=location.index + 1)] = occurrences.pop(location)
+
+        for subexpression in _subexpression_dfs(instruction):
+            self._usages[subexpression][CfgInstructionLocation(block, index)] += 1
```

Comment:
A unit-test is missing, you had a good test in the description, that you could implement to check that the algorithm works

Reply:
Added test with 55cdbd0